### DISCUSSION TITLE
Ears Style Guide

### DISCUSSION THREAD

**Discussion by holidaymike** [xmidt-org/ears#61] [General]
https://github.com/xmidt-org/ears/blob/main/docs/EarsStyleGuides.md

**Comment by holidaymike** [comment]
Some open questions with the initial draft:
* Do we want to style the file name? It seems that most files are in lower case with _test for test files. But I also see cases where people use camelCases.
* What's the thought on creating errors with errors.New(...). I have seen some golang standard library uses this approach and liked its simplicity:
```
SomethingIsWrong = errors.New("something is wrong")

return SomethingIsWrong

//checking
if err == SomethingIsWrong
```

**Reply by gtrevg** [reply]
```
SomethingIsWrong = errors.New("something is wrong")
```

This sentinel value pattern has been deprecated with the new errors functionality.

**Reply by gtrevg** [reply]
I do think we want to style the file names.  When you don't, I think things start to become an annoyance.  For instance (I tend to browse big projects and try to get a sense of what they do and how things play out):

https://github.com/kubernetes/kubernetes/tree/master/cmd/kubeadm/app/phases/selfhosting

![image](https://user-images.githubusercontent.com/1063149/102394478-cc068600-3fa7-11eb-9ad9-a0f518062a1b.png)

Some inconsistencies:
* should it be `pod_spec_mutation.go`?
* should it be `self_hosting_volumes.go`?

And then when we look at the includes:

```
package selfhosting

import (
	"fmt"
	"io/ioutil"
	"path/filepath"
	"strings"

	"k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	clientset "k8s.io/client-go/kubernetes"
	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
	"k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
)
```

We start seeing how file name choices impact includes and package definitions.  This is also directed by other conventions stated in Go with regard to [package names](https://blog.golang.org/package-names):

> They are lower case, with no under_scores or mixedCaps. They are often simple nouns


So if we start with package naming convention, we constrain the containing folder to the same naming convention -- folder names should be all lowercase (e.g.: `apiclient`).  That says nothing about the files contained in that package though.

If we use the base example of "a single file in a package that implements the package", in the `apiclient` example above, we would expect the single file in that package to be `apiclient.go`.  This would signify an all lowercase naming convention for all files within the package.

If we introduce testing, though, the convention is no longer maintained:
* Test packages are usually named something like `apiclient_test`
* Test files are usually named something like `apiclient_test.go`

At this point, we already have a mixed convention.  Because `_test` is core to how Go works, we do have to comply with that part of the convention.  I like mixed case over all lowercase in general as it makes things easier to read.  Both `apiclient_test` and `apiClient_test` seem odd, though, due to the mixed naming convention.  What about a third option for the files within a package?

```
api_client.go
api_client_test.go
```

With a snake_case convention, the file names are quickly read and it goes with the `_test` convention.

I did use that convention in a project before.  This is what it ended up looking like:

![image](https://user-images.githubusercontent.com/1063149/102398758-b300d380-3fad-11eb-9240-d92bbbfcf988.png)

It did seem easier to be able to parse what's going on and how files were grouped in purpose.  It seemed to work well with having the `_test.go` suffix as well.

What do you all think?
